Add upperBound, lowerBound, and equalRange

schmee opened the pull request:
This continues the work started in https://github.com/ziglang/zig/pull/15440 by @CraigglesO. The first commit is a squash of that PR, with attribution retained, and the second commit contains my own changes. From the message of that commit:

The old definitions had some problems:

- In `lowerBound`, the `lhs` (left hand side) argument was passed on the right hand side.
- In `upperBound`, the `greaterThan` function needed to return `greaterThanOrEqual` for the function work, so either the name or the implementation is incorrect.

To fix both problems, define the functions in terms of a `lessThan` function that returns `lhs < rhs`.
The is more consistent with the rest of `sort.zig` and it's also how C++ implements lower/upperBound (1)(2).

(1) https://en.cppreference.com/w/cpp/algorithm/lower_bound
(2) https://en.cppreference.com/w/cpp/algorithm/upper_bound

- Rewrite doc comments for brevity.
- Add a couple of more test cases.

andrewrk commented on lib/std/sort.zig:
```diff
@@ -502,6 +502,257 @@ test "binarySearch" {
+    key: anytype,
+    items: []const T,
+    context: anytype,
+    comptime lessThan: fn (context: @TypeOf(context), key: @TypeOf(key), lhs: T) bool,
+    comptime greaterThan: fn (context: @TypeOf(context), key: @TypeOf(key), rhs: T) bool,
```
Instead of these, it should take a compare function like `binarySearch` does.

andrewrk commented on lib/std/sort.zig:
```diff
@@ -502,6 +502,257 @@ test "binarySearch" {
     );
 }
 
+/// Returns the index of the first element in `items` not less than `key`,
+/// or `items.len` if no such element exists.
```
This is an incorrect description of what this function does. It returns 0 if the key is less than all items, and items.len if the key is greater than all items.

andrewrk commented on lib/std/sort.zig:
```diff
@@ -502,6 +502,257 @@ test "binarySearch" {
+/// Returns a tuple of the lower and upper bounds of `key` in `items`.
+/// See `lowerBound` and `upperBound` respectively.
+/// O(log n) complexity.
+pub fn equalRange(
```
I think this name is misleading. I suggest 
```suggestion
pub fn binarySearchSpan(
```
and to change the doc comments to be more accurate and describe the edge cases.

schmee replied:
Will do on the doc comments 👍 

The name comes from C++ (1) and the following property:

```
const nums = [_]i32{ 2, 4, 8, 8, 8, 8, 15, 22 };
const x: i32 = 8;
const range = equalRange(i32, x, &nums, {}, S.lower_i32);
const all_equal = nums[range[0]..range[1]]; // { 8, 8, 8, 8 } - the range where all items are equal to `x`
```
Of course Zig is not C++, but I personally don't find the name misleading. Happy to change to `binarySearchSpan` if that is still your preference.

(1) https://en.cppreference.com/w/cpp/algorithm/equal_range

andrewrk replied:
Looking at the unit tests again, I see that it does in fact return a range with only equal elements, and an empty range if no elements are equal. I think "equal range" is fine.

andrewrk reviewed (changes requested):
Thanks. I'm on board with adding these functions but the APIs are not mergeable yet.

schmee commented on lib/std/sort.zig:
```diff
@@ -502,6 +502,257 @@ test "binarySearch" {
+/// Returns the index of the first element in `items` not less than `key`,
+/// or `items.len` if no such element exists.
+/// O(log n) complexity.
+pub fn lowerBound(
```
Can do 👍 `std.sort.isSorted` currently takes a `lessThan` function, should that be changed to take a `compareFn` as well? I can work around it by wrapping the `compareFn` into a `lessThan` but that seems a bit kludgy.

schmee commented:
@andrewrk thank you for the review! I believe I have addressed all your comments except the assertion that the array is sorted, the reason being that it would make these functions `O(n)`, somewhat defeating the purpose of using binary search. The ordering requirement is now pointed out in the doc comments so the user can choose to add the assertion themselves if they want to trade speed for extra safety.

andrewrk commented:
That is a good point - it's not appropriate to change the big O complexity of a function in order to make an assertion.

andrewrk commented on lib/std/sort.zig:
```diff
@@ -502,6 +509,267 @@ test "binarySearch" {
+
+    while (left < right) {
+        const mid = left + (right - left) / 2;
+        if (compareFn(context, items[mid], key) == .lt) {
```
Didn't this used to be a `lessThan` function before? By changing it to `compareFn`, now it does potentially 2 comparisons instead of 1 in each iteration.

It looks like all the functions could have simply taken a `lessThan` function, no?

schmee replied:
They did, but I changed them because your comment: https://github.com/ziglang/zig/pull/18718#discussion_r1468955495, I must have misunderstood! So they should all take `lessThan`? I realize I may have confused you because in my previous commit `equalRange` took two functions (`lessThan` and `greaterThan`) but that is not necessary and all three functions can work with the same `lessThan` function.

andrewrk replied:
Yes I think you made other improvements along the way that now reveal it can be the ideal case of only taking one `lessThan` function. Sorry for the confusion. It was not obvious to me that it would work this way at first.

1. Most preferred scenario: only `lessThan`
2. Acceptable: only `compare`
3. No good: both `lessThan` and `greaterThan`

schmee commented:
@andrewrk updated to use `lessThan` across the board and squashed the commits, no other changes 👍

andrewrk commented:
Thanks, nice work!

jiripospisil commented on lib/std/sort.zig:
```diff
@@ -502,6 +509,267 @@ test "binarySearch" {
+
+    while (left < right) {
+        const mid = left + (right - left) / 2;
+        if (lessThan(context, items[mid], key)) {
```
@schmee This looks wrong. The type of `items[mid]` is `T`, not `@TypeOf(key)`. I think this only passes the tests because none of them end up with different types here.

schmee replied:
Yup! I actually ran into this myself the last few days, if no one else makes a PR to fix it I'll get around to it eventually 👍
